Refactor Stage IC: Move COM and Cycle Variables and Logic from YAML Master Templates to Python Scripts

AntonMFernando-NOAA opened the pull request:
<!--
  *** PLEASE READ ***

  Any PRs not following this template will be closed.

  Please delete all these comments before submitting the PR.

  Please use a short (<60 char), descriptive title for the PR title above. It should complete the sentence "If merged, this PR will _____". Capitalize the first word and do not end with a period.

  No content should appear above the "Description" header.

  If this PR is not merge-ready (e.g. it depends on other PRs not yet merged), please mark it as draft until it is ready.

  PRs should meet these guidelines:
  - Each PR should address ONE topic and have an associated issue.
  - No hard-coded paths or personal directories.
  - No temporary or backup files should be committed (including logs).
  - Any code that you disabled by being commented out should be removed or reenabled.
-->
# Description
- Move Jinja logic, com, and cycle vars in stage_ic YAML into Python; simplify YAML to component includes and switches only.
- This PR:
1. Implements per‑member COM path computation for gfs, gefs_offline, gefs_rt, gcafs, and enkfgdas cases
2. Move logic, member com paths and general variables to `stage_ic.py`

  Resolves #3946 


# Type of change
- [ ] Bug fix (fixes something broken)
- [x] New feature (adds functionality)
- [ ] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
<!-- Choose YES or NO from each of the following and delete the other -->
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO
- Does this change require an update to any of the following submodules? NO (If YES, please add a link to any PRs that are pending.)
  - [ ] EMC verif-global <!-- NOAA-EMC/EMC_verif-global#1234 -->
  - [ ] GDAS <!-- NOAA-EMC/GDASApp#1234 -->
  - [ ] GFS-utils <!-- NOAA-EMC/gfs-utils#1234 -->
  - [ ] GSI <!-- NOAA-EMC/GSI#1234 -->
  - [ ] GSI-monitor <!-- NOAA-EMC/GSI-Monitor#1234 -->
  - [ ] GSI-utils <!-- NOAA-EMC/GSI-Utils#1234 -->
  - [ ] UFS-utils <!-- ufs-community/UFS_UTILS#1234 -->
  - [ ] UFS-weather-model <!-- ufs-community/ufs-weather-model#1234 -->
  - [ ] wxflow <!-- NOAA-EMC/wxflow#1234 -->

# How has this been tested?
- Test cases on WCOSS2 that get affected by these changes 

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have documented my code, including function, input, and output descriptions
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] This change is covered by an existing CI test or a new one has been added
- [ ] Any new scripts have been added to the .github/CODEOWNERS file with owners
- [ ] I have made corresponding changes to the system documentation if necessary

AntonMFernando-NOAA committed:
added ush\python\pygfs\task\strageic_archive.py

AntonMFernando-NOAA committed:
update scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
update parm\stage\master_gefs.yaml.j2

AntonMFernando-NOAA committed:
update scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
update scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
added calculate_member_variables method

AntonMFernando-NOAA committed:
added stage_ic methods

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
added dictionary of stage_ic vars to stage_ic.py

AntonMFernando-NOAA committed:
removed vars rrom stage master yamls

AntonMFernando-NOAA committed:
update scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
update scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
update scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
modified ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
modify ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
modified scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
modified stage.py

AntonMFernando-NOAA committed:
updated scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
typos

AntonMFernando-NOAA committed:
Merge branch 'enabler/stage_archive' of https://github.com/AntonMFernando-NOAA/global-workflow into enabler/stage_archive

AntonMFernando-NOAA committed:
modified ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
update scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
Merge branch 'enabler/stage_archive' of https://github.com/AntonMFernando-NOAA/global-workflow into enabler/stage_archive

AntonMFernando-NOAA committed:
update per member

AntonMFernando-NOAA committed:
update stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
updated scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
edited sub-master.yamls

AntonMFernando-NOAA committed:
modified dev\workflow\rocoto\gefs_tasks.py

AntonMFernando-NOAA committed:
update parm\stage\master_gefs_RT.yaml.j2

AntonMFernando-NOAA committed:
Merge branch 'enabler/stage_archive' into develop

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update RT files

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
Merge branch 'enabler/stage_archive' of https://github.com/AntonMFernando-NOAA/global-workflow into enabler/stage_archive

AntonMFernando-NOAA committed:
update scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
upadte exscript

AntonMFernando-NOAA committed:
update scripts\exglobal_stage_ic.py

AntonMFernando-NOAA committed:
enable gefs_RT test (revert once done)

AntonMFernando-NOAA committed:
near-real-time change

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
removed copy and added link

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update yamls for gcafs

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update enkfgdas vars

AntonMFernando-NOAA committed:
modify parm\stage\master_gcafs.yaml.j2

AntonMFernando-NOAA committed:
update parm\stage\master_gcafs.yaml.j2

AntonMFernando-NOAA committed:
update vars

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
added comments

AntonMFernando-NOAA committed:
update parm\stage\analysis.yaml.j2

AntonMFernando-NOAA committed:
modify ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
modifed parm\stage\ice.yaml.j2

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
Merge branch 'develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
modified ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
Merge branch 'enabler/stage_archive' of https://github.com/AntonMFernando-NOAA/global-workflow into enabler/stage_archive

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

DavidHuber-NOAA commented on parm/stage/master_gefs_RT.yaml.j2:
```diff
@@ -25,63 +25,7 @@
-{% endif %}
-
-{% if ENSMEM == 0 %}
+{% if memdir == 0 %}
```
I don't think this is needed.

DavidHuber-NOAA replied:
Is this needed?

AntonMFernando-NOAA replied:
@DavidHuber-NOAA Yes. This is needed. When memdir=0 control mem yamls run and memdir>0 perturbed member yamls run.

DavidHuber-NOAA commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -56,3 +125,318 @@ def execute_stage(self, stage_dict: Dict[str, Any]) -> None:
+        # assign rRUN to RUN
+        case_vars.rRUN = case_vars.RUN
+        # case specific rRUN
+        if case_vars.RUN in ['gfs', 'gcdas', 'enkfgdas']:
```
Doesn't `stage_ic` also run on `RUN=gdas`?

DavidHuber-NOAA reviewed (commented):
I like this overall design a lot better than the Jinja-based logic.

DavidHuber-NOAA commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -56,3 +125,318 @@ def execute_stage(self, stage_dict: Dict[str, Any]) -> None:
+        if cycle_vars.previous_cycle:
+            cycle_vars.previous_cycle_YMD = cycle_vars.previous_cycle.strftime("%Y%m%d")
+            cycle_vars.previous_cycle_HH = cycle_vars.previous_cycle.strftime("%H")
+            cycle_vars.m_index = cycle_vars.current_cycle.hour // 6
```
Should `6` be replaced by `assim_freq`?

DavidHuber-NOAA commented on parm/stage/analysis.yaml.j2:
```diff
@@ -1,15 +1,8 @@
-        {% set imem = mem - first_mem %}
-        {% set COMOUT_ATMOS_ANALYSIS_MEM = COMOUT_ATMOS_ANALYSIS_MEM_list[imem] %}
         {% if DO_JEDIATMVAR %}
         {% for itile in range(6) %}
```
Should probably be replaced by `num_tiles`.

DavidHuber-NOAA replied:
Nudge.

AntonMFernando-NOAA replied:
@DavidHuber-NOAA range(6) --> num_tiles?

AntonMFernando-NOAA committed:
Merge branch 'develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
modified stage_ic.py and ex-script

AntonMFernando-NOAA committed:
update ush\python\pygfs\task\stage_ic.py

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Merge remote-tracking branch 'origin/develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
update more sub-yamls

AntonMFernando-NOAA committed:
Merge remote-tracking branch 'origin/develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Merge remote-tracking branch 'origin/develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Merge branch 'enabler/stage_archive' of https://github.com/AntonMFernando-NOAA/global-workflow into enabler/stage_archive

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
rework methods

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
remove list vars

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
update gcafs vars

AntonMFernando-NOAA committed:
Merge develop into enabler/stage_archive - resolved submodule and file conflicts

AntonMFernando-NOAA committed:
update _RT yamls

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

DavidHuber-NOAA commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -56,3 +184,239 @@ def execute_stage(self, stage_dict: Dict[str, Any]) -> None:
+        self.task_config['COMOUT_OCEAN_RESTART_PREV_MEM'] = self._replace_template_vars(
+            getattr(self.task_config, 'COM_OCEAN_RESTART_TMPL', ''), previous_cycle_mem_dict)
+        self.task_config['COMOUT_OCEAN_ANALYSIS_MEM'] = self._replace_template_vars(
+            getattr(self.task_config, 'COM_OCEAN_ANALYSIS_TMPL', ''), current_cycle_mem_dict)
```
Suggest writing a function to create these `COMOUT` variables from templates and populate a single `path_config` dictionary (rather than adding these to the `task_config`).

Pseudocode:
```python
def paths_from_templates(template_tuple_list):
  # template_tuple: (COM path (e.g. COMOUT_MED_RESTART_PREV_MEM), template (e.g. COM_MED_RESTART_TMPL), cycle_dict (e.g. previous_cycle_mem_dict))
  self.path_config={}
  for template_tuple in template_tuple_list:
      self.path_config[template_tuple[0]] = self._replace_template_vars(self.task_config.get(template_tuple[1], ''), template_tuple[2])
```

DavidHuber-NOAA commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -56,3 +184,239 @@ def execute_stage(self, stage_dict: Dict[str, Any]) -> None:
+        replaced_com = template
+        for var, value in var_dict.items():
+            replaced_com = replaced_com.replace(var, value)
+        return replaced_com
```
Since you are using, e.g., `getattr(self.task_config, 'COM_ATMOS_INPUT_TMPL', '')` as the argument to this function, it should check for empty strings and raise a `ValueError` if it is empty.

DavidHuber-NOAA commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -56,3 +184,239 @@ def execute_stage(self, stage_dict: Dict[str, Any]) -> None:
+        """
+        self.calculate_member()
+        run = getattr(self.task_config, 'RUN', None)
+        for memdir in range(self.task_config.first_mem, self.task_config.last_mem + 1):
```
This loop can probably be moved up to `exglobal_stage_ic.py`.

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
Update stage_ic.py

AntonMFernando-NOAA committed:
Refactor staging process for members

Replaced execute_stage_member with calculate_member and updated the loop to stage ICs for each member.

AntonMFernando-NOAA committed:
Merge branch 'develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
change methods

AntonMFernando-NOAA committed:
typos

AntonMFernando-NOAA committed:
update stage_ic vars

AntonMFernando-NOAA committed:
update methods

AntonMFernando-NOAA committed:
update parm/stage/master_gefs_RT.yaml.j2

AntonMFernando-NOAA committed:
update parm/stage/master_gefs_RT.yaml.j2

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,485 @@
+        cycle_vars['half_window'] = half_window
+
+        # Current cycle variables
+        if self.task_config.current_cycle:
```
`task_config` always has a `current_cycle` and `previous_cycle` -- this `if` check is not necessary.  Same `if` check on line 107 is not necessary.

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,485 @@
+            if self.task_config.DOIAU and self.task_config.MODE == "cycled":
+                cycle_vars['model_start_date_current_cycle'] = self.task_config.current_cycle + timedelta(hours=-half_window)
+            else:
+                if self.task_config.get('REPLAY_ICS', False):
```
`REPLAY_ICS` option was removed from `develop` in a previous PR.  This should be removed.

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,485 @@
+            cycle_vars['m_prefix'] = cycle_vars['model_start_date_current_cycle'].strftime("%Y%m%d.%H0000")
+
+        # Previous cycle variables
+        if self.task_config.previous_cycle:
```
See above comment https://github.com/NOAA-EMC/global-workflow/pull/3984/files#r2546652904

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,485 @@
+        # Current cycle variables
+        if self.task_config.current_cycle:
+            cycle_vars['current_cycle_HH'] = self.task_config.current_cycle.strftime("%H")
+            cycle_vars['current_cycle_YMD'] = self.task_config.current_cycle.strftime("%Y%m%d")
```
```suggestion
            cycle_vars['current_cycle_YMD'] = to_YMD(self.task_config.current_cycle)
```

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,485 @@
+                else:
+                    cycle_vars['model_start_date_current_cycle'] = self.task_config.current_cycle
+
+            cycle_vars['m_prefix'] = cycle_vars['model_start_date_current_cycle'].strftime("%Y%m%d.%H0000")
```
```suggestion
            cycle_vars['m_prefix'] = to_fv3time(cycle_vars['model_start_date_current_cycle'])
```

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,485 @@
+        if self.task_config.previous_cycle:
+            previous_cycle_HH = self.task_config.previous_cycle.strftime("%H")
+            cycle_vars['m_index'] = self.task_config.current_cycle.hour // self.task_config.assim_freq
+            cycle_vars['p_prefix'] = self.task_config.previous_cycle.strftime("%Y%m%d.%H0000")
```
Use `to_fv3time`

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,485 @@
+            cycle_vars['m_index'] = self.task_config.current_cycle.hour // self.task_config.assim_freq
+            cycle_vars['p_prefix'] = self.task_config.previous_cycle.strftime("%Y%m%d.%H0000")
+            cycle_vars['previous_cycle_HH'] = previous_cycle_HH
+            cycle_vars['previous_cycle_YMD'] = self.task_config.previous_cycle.strftime("%Y%m%d")
```
Use `to_YMD`

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,485 @@
+            'current_cycle_dict': {
+                "${ROTDIR}": rotdir,
+                "${RUN}": run,
+                "${YMD}": self.task_config.current_cycle.strftime("%Y%m%d"),
```
Use `to_YMD`

aerorahul replied:
here and elsewhere. (`to_fv3time` where applicable)

aerorahul commented on scripts/exglobal_stage_ic.py:
```diff
@@ -12,36 +12,24 @@
+    stage_dict = stage.create_stage_dict()
+
+    # Determine member range
+    first_mem, last_mem = stage._select_first_and_final_member(stage_dict.RUN, stage_dict.NMEM_ENS)
```
This would be better served if the method returns a list of members.
Because, in NRT, for GEFS, the members are staggered and 18z cycle, first = 61, last = 10 spanning 1-10, 61-80.

aerorahul commented on scripts/exglobal_stage_ic.py:
```diff
@@ -12,36 +12,24 @@
+    first_mem, last_mem = stage._select_first_and_final_member(stage_dict.RUN, stage_dict.NMEM_ENS)
+
+    # Loop through members and stage ICs for each
+    for member in range(first_mem, last_mem + 1):
```
If a list of members is received from above, this line can become:
```python
for member in member_list:
```
where `member_list = [1, 2, 3, ... ]` or as appropriate

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,485 @@
+        Returns
+        -------
+        Tuple[int, int]
+            Tuple of (first_member, last_member) indices
```
Return a list instead of tuple here.
Rename the method to `get_member_list`

AntonMFernando-NOAA committed:
update methods

AntonMFernando-NOAA committed:
Merge remote-tracking branch 'origin/develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,497 @@
+        else:
+            return []
+
+    def _map_gefs_member_to_gfs(self, m_index: int, member: int) -> int:
```
place this inside `get_member_list`, and make `get_member_list`  a staticmethod.

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,497 @@
+        Dict[str, str]
+            Dictionary mapping COM path keys to resolved file paths
+        """
+        path_dict = {}
```
```suggestion
        def _replace_template_vars( ...)
            # place this method here instead of down below.  This will turn this method into a static method
        path_dict = {}
```

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,497 @@
+            ('COMOUT_WAVE_RESTART_PREV_MEM', 'COM_WAVE_RESTART_TMPL', previous_cycle_mem_dict),
+        )
+
+        return self._paths_from_templates(stage_dict, com_paths)
```
can you just return the `com_paths` tuple and leave the `self._paths_from_templates` to be performed outside?  
This comment is for all `_get_member_com_paths_` methods

AntonMFernando-NOAA committed:
update yamls

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,497 @@
+        if stage_dict.RUN == 'gefs':
+            gefstype = self.task_config.get('GEFSTYPE', None)
+            if gefstype == 'gefs-real-time':
+                return self._get_member_com_paths_gefs_rt(stage_dict, member)
```
If you take the suggestion from below, and this method returns the `com_paths` tuple, you can compose the `com_paths` from that tuple and return the realized paths from here.

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,497 @@
             raise FileNotFoundError(f"FATAL ERROR: The ROTDIR ({stage_dict.ROTDIR}) does not exist!")
 
-        # Add the os.path.exists function to the dict for yaml parsing
         stage_dict['path_exists'] = os.path.exists
```
This can be added to the `create_stage_dict` method instead of here.

AntonMFernando-NOAA committed:
update stage_ic.py

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,470 @@
+        else:
+            return []
+
+    def _paths_from_templates(self, stage_dict: AttrDict, com_path_tuples: Tuple[Tuple[str, str, Dict[str, Any]], ...]) -> Dict[str, str]:
```
```suggestion
    @staticmethod
    def _paths_from_templates(stage_dict: AttrDict, com_path_tuples: Tuple[Tuple[str, str, Dict[str, Any]], ...]) -> Dict[str, str]:
```
This method no longer needs `self``

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,470 @@
+                for gefs_member in range(1, nmem_ens + 1):
+                    gfs_member = cyc_ranges[m_index][gefs_member - 1]
+                    member_list.append(gfs_member)
+                return member_list
```
```suggestion
                cyc_ranges = {'00': list(range(1, 31)),
                                       '06': list(range(21, 51)),
                                       '12': list(range(41, 71)),
                                       '18': list(range(61, 81)) + list(range(1, 11))}
                member_list = [0] + cyc_ranges[f'{m_index:02}']
                return member_list
```

aerorahul commented on scripts/exglobal_stage_ic.py:
```diff
@@ -12,40 +12,29 @@
+        stage_dict.RUN,
+        stage_dict.NMEM_ENS,
+        m_index=stage_dict.get('m_index', 0),
+        gefstype=stage.task_config.get('GEFSTYPE', None)
```
Put this in `stage_dict` instead of repeating this in multiple places.

AntonMFernando-NOAA committed:
update parm/stage/ice.yaml.j2

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,470 @@
+        ------
+        ValueError
+            If RUN type is unknown or GEFSTYPE is invalid for GEFS runs
+        """
+
+        if stage_dict.RUN == 'gefs':
+            gefstype = self.task_config.get('GEFSTYPE', None)
+            if gefstype == 'gefs-real-time':
+                return self._paths_from_templates(stage_dict, self._get_member_com_paths_gefs_rt(stage_dict, member))
+            elif gefstype == 'gefs-offline':
+                return self._paths_from_templates(stage_dict, self._get_member_com_paths_gefs_offline(stage_dict, member))
+            else:
+                raise ValueError(f"Invalid GEFSTYPE '{gefstype}' for RUN 'gefs'.")
+        elif stage_dict.RUN in ('gcafs', 'enkfgdas', 'gcdas', 'gdas'):
+            return self._paths_from_templates(stage_dict, self._get_member_com_paths_gcafs(stage_dict, member))
+        elif stage_dict.RUN == 'gfs':
+            return self._paths_from_templates(stage_dict, self._get_member_com_paths_gfs(stage_dict, member))
+        else:
+            raise ValueError(f"Unknown RUN type: {stage_dict.RUN}")
```
```suggestion
        if stage_dict.RUN == 'gefs':
            gefstype = self.task_config.get('GEFSTYPE', None)
            if gefstype == 'gefs-real-time':
                com_path = self._get_member_com_paths_gefs_rt(stage_dict, member)
            elif gefstype == 'gefs-offline':
                com_path = self._get_member_com_paths_gefs_offline(stage_dict, member)
            else:
                raise ValueError(f"Invalid GEFSTYPE '{gefstype}' for RUN 'gefs'.")
        elif stage_dict.RUN in ('gcafs', 'enkfgdas', 'gcdas', 'gdas'):
            com_path = self._get_member_com_paths_gcafs(stage_dict, member)
        elif stage_dict.RUN == 'gfs':
            com_path = self._get_member_com_paths_gfs(stage_dict, member)
        else:
            raise ValueError(f"Unknown RUN type: {stage_dict.RUN}")
            
         # Since we are here, we must have a com_path template to process
         return self._paths_from_templates(stage_dict, com_path)
```

aerorahul commented on scripts/exglobal_stage_ic.py:
```diff
@@ -12,40 +12,29 @@
+        stage_dict.update(stage.get_member_com_paths(stage_dict, member))
+
+        # Execute staging
+        stage.execute_stage(stage_dict)
```
```suggestion
        # Get member-specific COM paths and merge into stage_dict
        member_com_paths = stage.get_member_com_paths(stage_dict, member)
        
        stage_mem_dict = stage_dict.deepcopy()  # to ensure stage_dict does not get modified with each member
        stage_mem_dict.update(member_com_paths)

        # Execute staging
        stage.execute_stage(stage_mem_dict)
```

AntonMFernando-NOAA committed:
update stage_ic scripts

AntonMFernando-NOAA committed:
update parm/stage/ocean_rerun.yaml.j2

AntonMFernando-NOAA committed:
update comments ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
remove if conditions ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
update parm/stage/analysis.yaml.j2

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,480 @@
+        """
+        super().__init__(config)
+
+    def _copy_base_config(self) -> Dict[str, Any]:
```
```suggestion
    @logit(logger)
    def _copy_base_config(self) -> Dict[str, Any]:
```

Add `logit` decorator to all methods to allow tracing and debugging.

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Update ush/python/pygfs/task/stage_ic.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

AntonMFernando-NOAA committed:
Merge branch 'develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Update submodules to match develop: gdas.cd @ 3e774e7, gfs_utils.fd @ 8e46d14

AntonMFernando-NOAA committed:
Update ufs_utils.fd submodule to match develop @ 4e93a9f

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,481 @@
+        return path_dict
+
+    @logit(logger)
+    def _get_member_com_paths_gfs(self, stage_dict: AttrDict, member: int) -> Tuple[Tuple[str, str, Dict[str, Any]], ...]:
```
```suggestion
   @staticmethod
    def _get_member_com_paths_gfs( stage_dict: AttrDict, member: int) -> Tuple[Tuple[str, str, Dict[str, Any]], ...]:
```
These and the following are staticmethods, they don't need `self`.

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
add logit decorator

AntonMFernando-NOAA committed:
removed self from methods

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,494 @@
+            cycle_vars['model_start_date_current_cycle'] = self.task_config.current_cycle + timedelta(hours=-half_window)
+        else:
+            if self.task_config.get('REPLAY_ICS', False):
+                cycle_vars['model_start_date_current_cycle'] = self.task_config.current_cycle + timedelta(hours=half_window)
```
```suggestion
            cycle_vars['model_start_date_current_cycle'] = add_to_datetime(self.task_config.current_cycle, to_timedelta(-half_window))
        else:
            if self.task_config.get('REPLAY_ICS', False):
                cycle_vars['model_start_date_current_cycle'] = add_to_datetime(self.task_config.current_cycle, to_timedelta(half_window))

aerorahul commented on ush/python/pygfs/task/stage_ic.py:
```diff
@@ -1,58 +1,494 @@
-from wxflow import FileHandler, Task, logit, parse_j2yaml
+from typing import Any, Dict, Tuple, Optional
+from datetime import timedelta
+from wxflow import FileHandler, Task, logit, parse_j2yaml, AttrDict, to_YMD, to_fv3time
```
```suggestion
from wxflow import FileHandler, Task, add_to_datetime, logit, parse_j2yaml, AttrDict, to_YMD, to_fv3time, to_timedelta
```

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
add member to stage_dict

aerorahul reviewed (dismissed):
looks good.
let me know when you are ready for this to be tested via CI (after you have tested this atleast on one platform fully)

AntonMFernando-NOAA committed:
fix logic error

AntonMFernando-NOAA committed:
add sfs logic

AntonMFernando-NOAA committed:
Merge develop branch into enabler/stage_archive

AntonMFernando-NOAA committed:
resolved conflict and merge

AntonMFernando-NOAA committed:
Revert "resolved conflict and merge"

This reverts commit 45489bc28e693ffc52f64dbfff1f77ae2c3c7165.

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
Update gdas.cd submodule to match develop @ 58da1e3b

- Syncing gdas.cd from 3e774e7b to 58da1e3b
- Aligns with develop branch baseline
- Includes GDASApp atmospheric JEDI variational DA with GSI EnKF support (#1973)

AntonMFernando-NOAA committed:
update scripts/exglobal_stage_ic.py

AntonMFernando-NOAA committed:
update scripts/exglobal_stage_ic.py

AntonMFernando-NOAA committed:
add better indentation to analysis.yaml

AntonMFernando-NOAA committed:
update parm/stage/analysis.yaml.j2

AntonMFernando-NOAA committed:
update parm/stage/analysis.yaml.j2

AntonMFernando-NOAA committed:
update ush/python/pygfs/task/stage_ic.py

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

DavidHuber-NOAA commented on parm/stage/atmosphere_cold.yaml.j2:
```diff
@@ -1,18 +1,10 @@
-        {% set COMOUT_ATMOS_INPUT_MEM = COMOUT_ATMOS_INPUT_MEM_list[imem] %}
-        - ["{{ ICSDIR }}/{{ COMOUT_ATMOS_INPUT_MEM | relpath(ROTDIR) }}/gfs_ctrl.nc", "{{ COMOUT_ATMOS_INPUT_MEM }}"]
+        - ["{{ ICSDIR }}/{{ COMIN_ATMOS_INPUT | relpath(ROTDIR) }}/gfs_ctrl.nc", "{{ COMOUT_ATMOS_INPUT }}"]
         {% for ftype in ["gfs_data", "sfc_data"] %}
         {% for ntile in range(1, ntiles + 1) %}
-        - ["{{ ICSDIR }}/{{ COMOUT_ATMOS_INPUT_MEM | relpath(ROTDIR) }}/{{ ftype }}.tile{{ ntile }}.nc", "{{ COMOUT_ATMOS_INPUT_MEM }}"]
+        - ["{{ ICSDIR }}/{{ COMIN_ATMOS_INPUT | relpath(ROTDIR) }}/{{ ftype }}.tile{{ ntile }}.nc", "{{ COMOUT_ATMOS_INPUT }}"]
```
Is the plan to change all `COMOUT` variables to `COMIN` when pointing to the `ICSDIR`? If so, there are several templates still using `COMOUT`. Since this isn't an operational job, I'm OK with either approach, but we should probably be consistent.

DavidHuber-NOAA reviewed (dismissed):
Looks good. Thanks @AntonMFernando-NOAA!

AntonMFernando-NOAA committed:
change COMIN to COMOUT

AntonMFernando-NOAA committed:
bug fix modulefiles/gw_setup.gaeac6.lua

DavidHuber-NOAA commented:
Launching CI on C6 and WCOSS2.

DavidHuber-NOAA commented:
C6 appears to still be having issues. Launching CI manually.

DavidHuber-NOAA commented:
@AntonMFernando-NOAA the GEFS and GCAFS staging jobs failed on WCOSS2. Can you please take a look? The `RUNTESTS` directory can be found here: `/lfs/h2/emc/ptmp/david.huber/rt_3984`

DavidHuber-NOAA commented on dev/ci/cases/pr/C48_S2SWA_gefs_RT.yaml:
```diff
@@ -9,20 +9,20 @@ experiment:
   - gaeac6
   - gaeac5
   - hercules
-  - wcoss2
```
@AntonMFernando-NOAA Should this test be re-disabled?

AntonMFernando-NOAA replied:
Done.

AntonMFernando-NOAA committed:
typos

DavidHuber-NOAA commented:
Continuing tests on WCOSS2.

AntonMFernando-NOAA committed:
Merge branch 'NOAA-EMC:develop' into enabler/stage_archive

AntonMFernando-NOAA committed:
Revert "typos"

This reverts commit e31019993149031380813d22244bca22733bde20.

AntonMFernando-NOAA committed:
Revert "change COMIN to COMOUT"

This reverts commit 27249a80dca6e3d0436f0f25f69f3c9f5cec69f4.

AntonMFernando-NOAA committed:
remove wcoss2 from _RT tests

DavidHuber-NOAA commented:
All tests passed on Gaea C6. WCOSS2 reported one error for the `sfs_fcst_mem000_seg2` job. This is from the log `/lfs/h2/emc/ptmp/david.huber/rt_3984/COMROOT/C96mx100_S2S_3984/logs/1994050100/sfs_fcst_mem000_seg2.log`:

```
nid002148.cactus.wcoss2.ncep.noaa.gov 30: 
FATAL from PE     0: NetCDF: Unknown file format: netcdf_file_open:MOM6_OUTPUT/ocean_geometry.nc

nid002148.cactus.wcoss2.ncep.noaa.gov 30: 
FATAL from PE     0: NetCDF: Unknown file format: netcdf_file_open:MOM6_OUTPUT/ocean_geometry.nc

nid002148.cactus.wcoss2.ncep.noaa.gov 30: Image              PC                Routine            Line        Source      
libpnetcdf.so.4.0  0000148C9DE6686A  tracebackqq_          Unknown  Unknown
sfs_model.x        0000000005986F2E  mpp_mod_mp_mpp_er          72  mpp_util_mpi.inc
sfs_model.x        000000000583C442  fms_io_utils_mod_         190  fms_io_utils.F90
sfs_model.x        0000000005887E93  netcdf_io_mod_mp_         386  netcdf_io.F90
sfs_model.x        00000000058ACD2E  netcdf_io_mod_mp_         665  netcdf_io.F90
sfs_model.x        000000000583CFA4  fms_netcdf_domain         388  fms_netcdf_domain_io.F90
sfs_model.x        0000000005220F83  mom_io_infra_mp_o         152  MOM_io_infra.F90
sfs_model.x        0000000004B63D98  mom_io_file_mp_op         947  MOM_io_file.F90
sfs_model.x        0000000004B3B2DD  mom_io_mp_create_         350  MOM_io.F90
sfs_model.x        0000000004C79183  mom_shared_initia        1425  MOM_shared_initialization.F90
sfs_model.x        00000000047B562C  mom_mp_initialize        3129  MOM.F90
sfs_model.x        000000000474DD66  mom_ocean_model_n         284  mom_ocean_model_nuopc.F90
sfs_model.x        000000000470FDA0  mom_cap_mod_mp_in         723  mom_cap.F90
sfs_model.x        0000000000E8F604  _ZN5ESMCI6FTable1        2187  ESMCI_FTable.C
sfs_model.x        0000000000E93C7A  ESMCI_FTableCallE         844  ESMCI_FTable.C
sfs_model.x        00000000014A621E  _ZN5ESMCI3VMK5ent        1286  ESMCI_VMKernel.C
sfs_model.x        000000000148EDAF  _ZN5ESMCI2VM5ente        1303  ESMCI_VM.C
sfs_model.x        0000000000E90F67  c_esmc_ftablecall        1001  ESMCI_FTable.C
sfs_model.x        0000000000A38FA0  esmf_compmod_mp_e        1271  ESMF_Comp.F90
sfs_model.x        00000000008396D6  esmf_gridcompmod_        1443  ESMF_GridComp.F90
sfs_model.x        0000000000460AE7  nuopc_driver_mp_l        2906  NUOPC_Driver.F90
sfs_model.x        000000000048908F  nuopc_driver_mp_i        1347  NUOPC_Driver.F90
sfs_model.x        0000000000493004  nuopc_driver_mp_i         484  NUOPC_Driver.F90
sfs_model.x        00000nid002148.cactus.wcoss2.ncep.noaa.gov 30: 00000E8F604  _ZN5ESMCI6FTable1        2187  ESMCI_FTable.C
sfs_model.x        0000000000E93C7A  ESMCI_FTableCallE         844  ESMCI_FTable.C
sfs_model.x        00000000014A602F  _ZN5ESMCI3VMK5ent        2557  ESMCI_VMKernel.C
sfs_model.x        000000000148EDAF  _ZN5ESMCI2VM5ente        1303  ESMCI_VM.C
sfs_model.x        0000000000E90F67  c_esmc_ftablecall        1001  ESMCI_FTable.C
sfs_model.x        0000000000A38FA0  esmf_compmod_mp_e        1271  ESMF_Comp.F90
sfs_model.x        00000000008396D6  esmf_gridcompmod_        1443  ESMF_GridComp.F90
sfs_model.x        0000000000431695  MAIN__                    392  UFS.F90
sfs_model.x        00000000004304D2  Unknown               Unknown  Unknown
libc-2.31.so       0000148C9B43E24D  __libc_start_main     Unknown  Unknown
sfs_model.x        00000000004303EA  Unknown               Unknown  Unknown
MPICH ERROR [Rank 30] [job id 80b6bde2-878c-4355-b888-4a2d965e38ec] [Wed Nov 26 22:32:19 2025] [nid002148] - Abort(1) (rank 30 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 30
```

This error was reported in one other PR, also on WCOSS2: https://github.com/NOAA-EMC/global-workflow/pull/3833#issuecomment-3089594754 and disappeared on rerun of the SFS test. This likely represents an intermittent netCDF issue. I will open a separate issue for this and mark the CI as passed on WCOSS2 for the current PR.

DavidHuber-NOAA commented:
I just relaunched the failed `sfs_fcst_mem000_seg2` job and it is now running successfully. I suspect a possible node issue, so I will hold off on opening an issue and instead inform the WCOSS2 helpdesk.

DavidHuber-NOAA commented:
After relaunching the SFS forecast, all jobs completed successfully on WCOSS2.  Merging.
